src/rabbit_vhost_dead_letter.erl

```diff
@@ -0,0 +1,135 @@
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at http://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2018 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_vhost_dead_letter).
+
+-include("rabbit.hrl").
+
+-behaviour(gen_server).
+
+-export([start/1, start_link/0]).
+-export([stop/1]).
+-export([publish/5]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
+         code_change/3]).
+
+-record(state, {queue_states,
+                queue_cleanup_timer}).
+
+start(VHost) ->
+    case rabbit_vhost_sup_sup:get_vhost_sup(VHost) of
+        {ok, VHostSup} ->
+            supervisor2:start_child(VHostSup,
+                                    {rabbit_vhost_dead_letter,
+                                     {rabbit_vhost_dead_letter, start_link, []},
+                                     transient, ?WORKER_WAIT, worker, [rabbit_vhost_dead_letter]});
+        {error, {no_such_vhost, VHost}} = E ->
+            rabbit_log:error("Failed to start a dead letter process for vhost ~s: vhost no"
+                             " longer exists!", [VHost]),
+            E
+    end.
+
+stop(VHost) ->
+    case rabbit_vhost_sup_sup:get_vhost_sup(VHost) of
+        {ok, VHostSup} ->
+            ok = supervisor2:terminate_child(VHostSup, rabbit_vhost_dead_letter),
+            ok = supervisor2:delete_child(VHostSup, rabbit_vhost_dead_letter);
+        {error, {no_such_vhost, VHost}} ->
+            rabbit_log:error("Failed to stop a dead letter process for vhost ~s: "
+                             "vhost no longer exists!", [VHost]),
+
+            ok
+    end.
+
+publish(VHost, X, RK, QName, ReasonMsgs) ->
+    case vhost_dead_letter_pid(VHost) of
+        no_pid ->
+            %% TODO what to do???
```

Comment:
Should this TODO be addressed? If vhost does not have this process, does it mean that the vhost supervisor should restart?

Reply:
I don't know if there is much we can do, except maybe to retry. The supervisor should restart the crashed process, so if that's not happening it's because something went terribly wrong. I don't think crashing here would be an acceptable behaviour, as the delivery itself is a cast (https://github.com/rabbitmq/rabbitmq-server/blob/master/src/rabbit_amqqueue.erl#L1242). Anything that went wrong there we wouldn't know anyway.